src/InodeCache.cpp

```diff
@@ -0,0 +1,528 @@
+    fprintf(stderr,
+            "Failed to allocate file space for inode cache: %s\n",
+            strerror(errno));
+    free(buf);
+    close(fd);
+    return false;
+  }
+  free(buf);
+#  endif
+  SharedRegion* sr = reinterpret_cast<SharedRegion*>(mmap(
+    nullptr, sizeof(SharedRegion), PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0));
+  if (sr == reinterpret_cast<void*>(-1)) {
+    fprintf(stderr, "Failed to mmap new inode cache: %s\n", strerror(errno));
+    close(fd);
+    return false;
+  }
+
+  // Initialize new shared region.
+  sr->version = k_version;
+  pthread_mutexattr_t mattr;
+  pthread_mutexattr_init(&mattr);
+  pthread_mutexattr_setpshared(&mattr, PTHREAD_PROCESS_SHARED);
+#  ifdef PTHREAD_MUTEX_ROBUST
+  pthread_mutexattr_setrobust(&mattr, PTHREAD_MUTEX_ROBUST);
+#  endif
+  for (uint32_t i = 0; i < k_num_buckets; ++i)
+    pthread_mutex_init(&sr->buckets[i].mt, &mattr);
+
+  munmap(sr, sizeof(SharedRegion));
+
+  // linkat() will fail silently if a file with the same name already exists.
+  // This will be the case if two processes tries to create a new file
+  // simultaneously. Thus close the current file handle and reopen a new one,
+  // which will make us using the first created file also if we didn't win the
+  // race.
+  if (linkat(
+        AT_FDCWD, path_buf, AT_FDCWD, filename.c_str(), AT_SYMLINK_FOLLOW)) {
+    cc_log("Failed to link new inode cache: %s", strerror(errno));
+#  ifndef O_TMPFILE
+    unlink(path_buf);
+#  endif
+    close(fd);
+    return false;
+  }
+
+#  ifndef O_TMPFILE
+  unlink(path_buf);
+#  endif
+  close(fd);
+  return true;
+}
+
+bool
+initialize(const Config& config)
+{
+  if (!config.inode_cache())
+    return false;
+
+  if (g_sr)
+    return true;
+
+  std::string filename = get_file_from_config(config);
+  if (g_sr || mmap_file(filename))
+    return true;
+
+  // Try to create a new cache if we failed to map an existing file.
+  create_new_file(filename);
+
+  // Concurrent processes could try to create new files simultaneously and the
+  // file that actually landed on disk will be from the process that won the
+  // race. Thus we try to open the file from disk instead of reusing the file
+  // handle to the file we just created.
+  return mmap_file(filename);
+}
+
+} // namespace
+
+bool
+get(const Config& config,
+    const char* path,
+    digest* file_digest,
+    int* return_value)
+{
+  if (!initialize(config))
+    return false;
+
+  digest key_digest;
+  if (!hash_inode(config, path, &key_digest))
+    return false;
+
+  Bucket* bucket = acquire_bucket(key_digest);
+
+  if (!bucket)
+    return false;
+
+  for (uint32_t i = 0; i < k_num_entries; ++i) {
+    if (digests_equal(&bucket->entries[i].key_digest, &key_digest)) {
+      if (i) {
+        Entry tmp;
+        memcpy(&tmp, &bucket->entries[i], sizeof(Entry));
+        memmove(&bucket->entries[1], &bucket->entries[0], sizeof(Entry) * i);
+        memcpy(&bucket->entries[0], &tmp, sizeof(Entry));
```

Comment:
Could this be simplified to

```suggestion
        Entry tmp = bucket->entries[i];
        memmove(&bucket->entries[1], &bucket->entries[0], sizeof(Entry) * i);
        bucket->entries[0] = tmp;
```

?